libyara/base64.c

```diff
@@ -0,0 +1,495 @@
+    }
+    printf("\n");
+
+    p = p->next;
+  }
+
+  return;
+}
+
+
+//
+// Destroy a list of base64 nodes.
+//
+static void _yr_base64_destroy_nodes(
+    BASE64_NODE* head)
+{
+  BASE64_NODE* p = head;
+  BASE64_NODE* next;
+
+  while (p != NULL)
+  {
+    yr_free(p->str);
+    next = p->next;
+    yr_free(p);
+    p = next;
+  }
+
+  return;
+}
+
+
+//
+// Create the regexp that is the alternatives of each of the strings collected
+// in the BASE64_NODE list.
+//
+int _yr_base64_create_regexp(
+    BASE64_NODE* head,
+    RE_AST** re_ast,
+    RE_ERROR* re_error)
+{
+  char* re_str;
+  char* s;
+  uint32_t i;
+  uint32_t length = 0;
+  uint32_t c = 0; // The number of nodes in the list, used to know how many '|'.
+  BASE64_NODE* p = head;
+
+  while (p != NULL)
+  {
+    length += (p->str->length + p->escaped);
+    c++;
+    p = p->next;
+  }
+
+  if (c == 0)
+    return ERROR_INSUFFICIENT_MEMORY;
+
+  // Make sure to include '(' and ')'.
+  // The number of '|' is number of nodes - 1.
+  re_str = (char*) yr_malloc(length + 2 + (c - 1) + 1);
+  if (re_str == NULL)
+    return ERROR_INSUFFICIENT_MEMORY;
+
+  s = re_str;
+  p = head;
+  *s++ = '(';
+  while (p != NULL)
+  {
+    for (i = 0; i < p->str->length; i++)
+    {
+      if (IS_METACHAR(p->str->c_string[i]))
+        *s++ = '\\';
+
+      if (p->str->c_string[i] == '\x00')
+      {
+        *s++ = '\\';
+        *s++ = 'x';
+        *s++ = '0';
+        *s++ = '0';
+      }
+      else
+        *s++ = p->str->c_string[i];
+    }
+
+    if (p->next != NULL)
+      *s++ = '|';
+
+    p = p->next;
+  }
+  *s++ = ')';
+  *s = '\x00';
+
+  // Useful for debugging as long as the string has no NULL bytes in it. ;)
+  //printf("%s\n", re_str);
+
+  FAIL_ON_ERROR_WITH_CLEANUP(
+      yr_re_parse(re_str, re_ast, re_error),
+      yr_free(re_str));
+
+  yr_free(re_str);
+  return ERROR_SUCCESS;
+}
+
+
+//
+// Given a string and an alphabet, generate the RE_AST suitable for representing
+// the different encodings of the string. This means we generate
+// "(ABCD|EFGH|IJKL)" and must be careful to escape any special characters as
+// a result of the base64 encoding.
+//
+// This uses ideas from:
+// https://www.leeholmes.com/blog/2019/12/10/searching-for-content-in-base-64-strings-2/
+//
+// This does not emit the code for the RE. A further call to yr_re_ast_emit_code
+// is required to get the code.
+//
+int yr_base64_ast_from_string(
+    SIZED_STRING* in_str,
+    YR_MODIFIER modifier,
+    RE_AST** re_ast,
+    RE_ERROR* error)
+{
+  BASE64_NODE* head = NULL;
+  BASE64_NODE* tail = NULL;
+  SIZED_STRING* wide_str;
+
+  if (modifier.flags & STRING_GFLAGS_WIDE)
+  {
+    wide_str = sized_string_convert_to_wide(in_str);
+
+    if (modifier.flags & STRING_GFLAGS_BASE64)
+    {
+      FAIL_ON_ERROR_WITH_CLEANUP(
+          _yr_base64_create_nodes(wide_str, modifier.alphabet, 0, &head, &tail),
+          {
+            strcpy(error->message, "Failure encoding base64 wide string");
```

Comment:
Are these errors necessary? I mean, if `_yr_base64_create_nodes` fails only on out-of-memory conditions, we don't need to add an error message here, and this function doesn't need to receive the pointer to `RE_ERROR`.

Reply:
I did it this way so it matches `yr_re_parse_hex()` and `yr_re_parse()` when called from parser.c. Also, I eventually call into `yr_re_parse()` to do the final parsing, which requires an `RE_ERROR*`. This is arguably the more important piece since if the RE parsing fails we would like to get detailed error information.